bug: vite 5.3.0 fails ionic build command

### Describe the bug

running ionic build with vite 5.3.0 gives error, 5.2.11 works.

### Reproduction

......

### Steps to reproduce

in package.json use vite: "vite": "5.3.0"

```
sil@SIL-IMAC-CH mxmobile % ionic build
> vite build
vite v5.3.0 building for production...
✓ 71 modules transformed.
x Build failed in 374ms
error during build:
[commonjs--resolver] node_modules/@ionic/core/components/ion-item-sliding.js (53:78): Expected ',', got '('
file: /Users/sil/Local/Weinberg/_code/mxmobile/node_modules/@ionic/core/components/ion-item-sliding.js:53:78

51:         });
52:         await this.updateOptions();
53:         this.gesture = (await import('./index3.js')).createGesture({
                                                                        ^
54:             el,
55:             gestureName: 'item-swipe',

RollupError: Expected ',', got '('
    at getRollupError (file:///Users/sil/Local/Weinberg/_code/mxmobile/node_modules/rollup/dist/es/shared/parseAst.js:396:41)
    at ParseError.initialise (file:///Users/sil/Local/Weinberg/_code/mxmobile/node_modules/rollup/dist/es/shared/node-entry.js:11332:28)
    at convertNode (file:///Users/sil/Local/Weinberg/_code/mxmobile/node_modules/rollup/dist/es/shared/node-entry.js:13082:10)
    at convertProgram (file:///Users/sil/Local/Weinberg/_code/mxmobile/node_modules/rollup/dist/es/shared/node-entry.js:12399:12)
    at Module.setSource (file:///Users/sil/Local/Weinberg/_code/mxmobile/node_modules/rollup/dist/es/shared/node-entry.js:14246:24)
    at async ModuleLoader.addModuleSource (file:///Users/sil/Local/Weinberg/_code/mxmobile/node_modules/rollup/dist/es/shared/node-entry.js:18892:13)
[ERROR] An error occurred while running subprocess vite.
        
        vite build exited with exit code 1.
        
        Re-running this command with the --verbose flag may provide more information.
sil@SIL-IMAC-CH mxmobile % 
```

and now with vite 5.2.11

```
sil@SIL-IMAC-CH mxmobile % ionic build
> vite build
vite v5.2.11 building for production...
✓ 1103 modules transformed.
dist/assets/web-legacy-Dv5WM5Nk.js                 0.38 kB │ gzip:   0.27 kB
dist/assets/status-tap-legacy-BB5Xm_Qf.js          0.60 kB │ gzip:   0.40 kB
dist/assets/swipe-back-legacy-j1_b47wM.js          0.78 kB │ gzip:   0.52 kB
dist/assets/md.transition-legacy-mc4ssSAF.js       1.13 kB │ gzip:   0.62 kB
dist/assets/index9-legacy-BwuAEZsE.js              1.74 kB │ gzip:   0.88 kB
dist/assets/input-shims-legacy-cWoHyAkS.js         4.73 kB │ gzip:   2.08 kB
dist/assets/web-legacy-D1JDicTQ.js                 8.70 kB │ gzip:   2.91 kB
dist/assets/ios.transition-legacy-CWj5cWCJ.js     10.64 kB │ gzip:   3.03 kB
dist/assets/polyfills-legacy-C7qq4mXn.js          66.71 kB │ gzip:  26.81 kB
dist/assets/index-legacy-B6OrpkwG.js             199.19 kB │ gzip:  60.29 kB
dist/assets/xlsx-legacy-bbAQ7zft.js              459.96 kB │ gzip: 149.82 kB
dist/assets/vfs_fonts-legacy-CTS0afYD.js         870.31 kB │ gzip: 427.21 kB
dist/assets/pdfmake-legacy-DjJJWIxI.js         1,175.30 kB │ gzip: 530.75 kB
dist/assets/index-legacy-BjFH0Qow.js           2,040.85 kB │ gzip: 463.19 kB

(!) Some chunks are larger than 500 kB after minification. Consider:
- Using dynamic import() to code-split the application
- Use build.rollupOptions.output.manualChunks to improve chunking: https://rollupjs.org/configuration-options/#output-manualchunks
- Adjust chunk size limit for this warning via build.chunkSizeWarningLimit.
dist/index.html                             2.33 kB │ gzip:   1.03 kB
dist/assets/index-CEdewMNL.css             34.13 kB │ gzip:   5.70 kB
dist/assets/web-CgTZG40O.js                 0.29 kB │ gzip:   0.21 kB
dist/assets/status-tap-D5_DpQQP.js          0.48 kB │ gzip:   0.34 kB
dist/assets/swipe-back-Btt47J0Y.js          0.68 kB │ gzip:   0.47 kB
dist/assets/md.transition-BUov7XLQ.js       1.07 kB │ gzip:   0.57 kB
dist/assets/index9-BSBzLANq.js              1.63 kB │ gzip:   0.83 kB
dist/assets/input-shims-CtaETH8b.js         4.82 kB │ gzip:   2.08 kB
dist/assets/web-DeNFjlLC.js                 8.68 kB │ gzip:   2.96 kB
dist/assets/ios.transition-DxDFfbv7.js     11.21 kB │ gzip:   3.10 kB
dist/assets/index-BxYELR8Z.js             198.77 kB │ gzip:  62.14 kB
dist/assets/xlsx-BP7NNpC4.js              470.76 kB │ gzip: 153.14 kB
dist/assets/vfs_fonts-BbT5ejxm.js         870.24 kB │ gzip: 427.14 kB
dist/assets/pdfmake-UpCaLPvY.js         1,206.83 kB │ gzip: 548.63 kB
dist/assets/index-D30E46v-.js           2,024.71 kB │ gzip: 468.33 kB

(!) Some chunks are larger than 500 kB after minification. Consider:
- Using dynamic import() to code-split the application
- Use build.rollupOptions.output.manualChunks to improve chunking: https://rollupjs.org/configuration-options/#output-manualchunks
- Adjust chunk size limit for this warning via build.chunkSizeWarningLimit.
✓ built in 29.26s
sil@SIL-IMAC-CH mxmobile % 
```



### System Info

```shell
System:
    OS: macOS 14.5
    CPU: (12) x64 Intel(R) Core(TM) i5-10500 CPU @ 3.10GHz
    Memory: 9.61 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.13.1 - /usr/local/opt/node@20/bin/node
    npm: 10.5.2 - /usr/local/opt/node@20/bin/npm
  Browsers:
    Chrome: 125.0.6422.142
    Safari: 17.5
  npmPackages:
    @vitejs/plugin-legacy: ^5.0.0 => 5.4.1 
    @vitejs/plugin-react: ^4.0.1 => 4.3.1 
    vite: 5.2.11 => 5.2.11
```


### Used Package Manager

npm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/plugins/importAnalysisBuild.ts
+++ b/packages/vite/src/node/plugins/importAnalysisBuild.ts
@@ -42,7 +42,7 @@ const preloadMarkerRE = new RegExp(preloadMarker, 'g')
 const dynamicImportPrefixRE = /import\s*\(/
 
 const dynamicImportTreeshakenRE =
-  /(\b(const|let|var)\s+(\{[^}.]+\})\s*=\s*await\s+import\([^)]+\))|(\(\s*await\s+import\([^)]+\)\s*\)(\??\.[^;[\s]+)+)|\bimport\([^)]+\)(\s*\.then\([^{]*?\(\s*\{([^}.]+)\})/g
+  /((?:\bconst\s+|\blet\s+|\bvar\s+|,\s*)(\{[^}.]+\})\s*=\s*await\s+import\([^)]+\))|(\(\s*await\s+import\([^)]+\)\s*\)(\??\.[\w$]+))|\bimport\([^)]+\)(\s*\.then\([^{]*?\(\s*\{([^}.]+)\})/g
 
 function toRelativePath(filename: string, importer: string) {
   const relPath = path.posix.relative(path.posix.dirname(importer), filename)
@@ -252,48 +252,47 @@ export function buildImportAnalysisPlugin(config: ResolvedConfig): Plugin {
           /* handle `const {foo} = await import('foo')`
            *
            * match[1]: `const {foo} = await import('foo')`
-           * match[2]: `const`
-           * match[3]: `{foo}`
+           * match[2]: `{foo}`
            * import end: `const {foo} = await import('foo')_`
            *                                               ^
            */
           if (match[1]) {
             dynamicImports[dynamicImportTreeshakenRE.lastIndex] = {
-              declaration: `${match[2]} ${match[3]}`,
-              names: match[3]?.trim(),
+              declaration: `const ${match[2]}`,
+              names: match[2]?.trim(),
             }
             continue
           }
 
           /* handle `(await import('foo')).foo`
            *
-           * match[4]: `(await import('foo')).foo`
-           * match[5]: `.foo`
+           * match[3]: `(await import('foo')).foo`
+           * match[4]: `.foo`
            * import end: `(await import('foo'))`
            *                                  ^
            */
-          if (match[4]) {
-            let names = match[5].match(/\.([^.?]+)/)?.[1] || ''
+          if (match[3]) {
+            let names = match[4].match(/\.([^.?]+)/)?.[1] || ''
             // avoid `default` keyword error
             if (names === 'default') {
               names = 'default: __vite_default__'
             }
             dynamicImports[
-              dynamicImportTreeshakenRE.lastIndex - match[5]?.length - 1
+              dynamicImportTreeshakenRE.lastIndex - match[4]?.length - 1
             ] = { declaration: `const {${names}}`, names: `{ ${names} }` }
             continue
           }
 
           /* handle `import('foo').then(({foo})=>{})`
            *
-           * match[6]: `.then(({foo}`
-           * match[7]: `foo`
+           * match[5]: `.then(({foo}`
+           * match[6]: `foo`
            * import end: `import('foo').`
            *                           ^
            */
-          const names = match[7]?.trim()
+          const names = match[6]?.trim()
           dynamicImports[
-            dynamicImportTreeshakenRE.lastIndex - match[6]?.length
+            dynamicImportTreeshakenRE.lastIndex - match[5]?.length
           ] = { declaration: `const {${names}}`, names: `{ ${names} }` }
         }
       }

--- a/playground/dynamic-import/__tests__/dynamic-import.spec.ts
+++ b/playground/dynamic-import/__tests__/dynamic-import.spec.ts
@@ -193,6 +193,12 @@ test('dynamic import treeshaken log', async () => {
   expect(log).not.toContain('treeshaken removed')
 })
 
+test('dynamic import syntax parsing', async () => {
+  const log = browserLogs.join('\n')
+  expect(log).toContain('treeshaken syntax foo')
+  expect(log).toContain('treeshaken syntax default')
+})
+
 test.runIf(isBuild)('dynamic import treeshaken file', async () => {
   expect(findAssetFile(/treeshaken.+\.js$/)).not.toContain('treeshaken removed')
 })

--- a/playground/dynamic-import/nested/index.js
+++ b/playground/dynamic-import/nested/index.js
@@ -138,6 +138,7 @@ import(`../nested/nested/${base}.js`).then((mod) => {
 ;(async function () {
   const { foo } = await import('./treeshaken/treeshaken.js')
   const { bar, default: tree } = await import('./treeshaken/treeshaken.js')
+  const default2 = (await import('./treeshaken/treeshaken.js')).default
   const baz1 = (await import('./treeshaken/treeshaken.js')).baz1
   const baz2 = (await import('./treeshaken/treeshaken.js')).baz2.log
   const baz3 = (await import('./treeshaken/treeshaken.js')).baz3?.log
@@ -145,18 +146,35 @@ import(`../nested/nested/${base}.js`).then((mod) => {
     ({ baz4 }) => baz4,
   )
   const baz5 = await import('./treeshaken/treeshaken.js').then(function ({
-    baz5,
-  }) {
-    return baz5
-  })
+      baz5,
+    }) {
+      return baz5
+    }),
+    { baz6 } = await import('./treeshaken/treeshaken.js')
   foo()
   bar()
   tree()
+  ;(await import('./treeshaken/treeshaken.js')).default()
+  default2()
   baz1()
   baz2()
   baz3()
   baz4()
   baz5()
+  baz6()
+})()
+// Test syntax parsing only
+;(async function () {
+  const default1 = await import('./treeshaken/syntax.js').then(
+    (mod) => mod.default,
+  )
+  const default2 = (await import('./treeshaken/syntax.js')).default,
+    other = () => {}
+  const foo = await import('./treeshaken/syntax.js').then((mod) => mod.foo)
+  default1()
+  default2()
+  other()
+  foo()
 })()
 
 import(`../nested/static.js`).then((mod) => {

--- a/playground/dynamic-import/nested/treeshaken/syntax.js
+++ b/playground/dynamic-import/nested/treeshaken/syntax.js
@@ -1,0 +1,6 @@
+export const foo = () => {
+  console.log('treeshaken syntax foo')
+}
+export default () => {
+  console.log('treeshaken syntax default')
+}

--- a/playground/dynamic-import/nested/treeshaken/treeshaken.js
+++ b/playground/dynamic-import/nested/treeshaken/treeshaken.js
@@ -23,6 +23,9 @@ export const baz4 = () => {
 export const baz5 = () => {
   console.log('treeshaken baz5')
 }
+export const baz6 = () => {
+  console.log('treeshaken baz6')
+}
 export const removed = () => {
   console.log('treeshaken removed')
 }

